source/slang/source-loc.cpp

```diff
@@ -5,30 +5,261 @@
 
 namespace Slang {
 
-String ExpandedSourceLoc::getPath() const
+/* !!!!!!!!!!!!!!!!!!!!!!!!! SourceUnit !!!!!!!!!!!!!!!!!!!!!!!!!!!! */
+
+int SourceUnit::findEntryIndex(SourceLoc sourceLoc) const
 {
-    if(!sourceManager)
-        return String();
+    if (!m_range.contains(sourceLoc))
+    {
+        return -1;
+    }
+
+    const auto rawValue = sourceLoc.getRaw();
+
+    int hi = int(m_entries.Count());    
+    // If there are no entries, or it is in front of the first entry, then there is no associated entry
+    if (hi == 0 || 
+        m_entries[0].m_startLoc.getRaw() > sourceLoc.getRaw())
+    {
+        return -1;
+    }
+
+    int lo = 0;
+    while (lo + 1 < hi)
+    {
+        int mid = (hi + lo) >> 1;
+
+        const Entry& midEntry = m_entries[mid];
+        SourceLoc::RawValue midValue = midEntry.m_startLoc.getRaw();
 
-    return sourceManager->sourceFiles[entryIndex].path;
+        if (midValue <= rawValue)
+        {
+            // The location we seek is at or after this entry
+            lo = mid;
+        }
+        else
+        {
+            // The location we seek is before this entry
+            hi = mid;
+        }
+    }
+
+    return lo;
 }
 
-String ExpandedSourceLoc::getSpellingPath() const
+void SourceUnit::addLineDirective(SourceLoc directiveLoc, StringSlicePool::Handle pathHandle, int line)
 {
-    if(!sourceManager)
-        return String();
+    SLANG_ASSERT(pathHandle != StringSlicePool::Handle(0));
+    SLANG_ASSERT(m_range.contains(directiveLoc));
+
+    // Check that the directiveLoc values are always increasing
+    SLANG_ASSERT(m_entries.Count() == 0 || (m_entries.Last().m_startLoc.getRaw() < directiveLoc.getRaw()));
 
-    return sourceManager->sourceFiles[entryIndex].sourceFile->path;
+    // Calculate the offset
+    const int offset = m_range.getOffset(directiveLoc);
+    
+    // Get the line index in the original file
+    const int lineIndex = m_sourceFile->calcLineIndexFromOffset(offset);
+
+    Entry entry;
+    entry.m_startLoc = directiveLoc;
+    entry.m_pathHandle = pathHandle;
+    
+    // We also need to make sure that any lookups for line numbers will
+    // get corrected based on this files location.
+    // We assume the line number coming in is a line number, NOT an index so the correction needs + 1
+    // There is an additional + 1 because we want the NEXT line to be 99 (ie +2 is correct 'fix')
```

Comment:
>  because we want the NEXT line to be 99

Is there a missing example that should be in the comment, something like:

```hlsl
#line 99
void thisFuncShouldBeOnLine99() {}
```
?